tests/integration/functional/zen_stores/test_zen_store.py

```diff
@@ -0,0 +1,991 @@
+        )
+
+    default_orchestrator.name = "axl"
+    with pytest.raises(IllegalOperationError):
+        store.update_stack_component(
+            component_id=default_artifact_store.id,
+            component_update=component_update,
+        )
+
+
+def test_delete_default_stack_component_fails():
+    """Tests that deleting default stack components is prohibited."""
+    client = Client()
+    store = client.zen_store
+    default_artifact_store = store.list_stack_components(
+        ComponentFilterModel(
+            workspace_id=client.active_workspace.id,
+            type=StackComponentType.ARTIFACT_STORE,
+            name="default",
+        )
+    )[0]
+
+    default_orchestrator = store.list_stack_components(
+        ComponentFilterModel(
+            workspace_id=client.active_workspace.id,
+            type=StackComponentType.ORCHESTRATOR,
+            name="default",
+        )
+    )[0]
+
+    with pytest.raises(IllegalOperationError):
+        store.delete_stack_component(default_artifact_store.id)
+
+    with pytest.raises(IllegalOperationError):
+        store.delete_stack_component(default_orchestrator.id)
+
+
+# .-------------------------.
+# | Stack component flavors |
+# '-------------------------'
+
+# .--------.
+# | STACKS |
+# '--------'
+
+
+def test_updating_default_stack_fails():
+    """Tests that updating the default stack is prohibited."""
+    client = Client()
+
+    default_stack = client.get_stack(DEFAULT_STACK_NAME)
+    assert default_stack.name == DEFAULT_WORKSPACE_NAME
+    stack_update = StackUpdateModel(name="axls_stack")
+    with pytest.raises(IllegalOperationError):
+        client.zen_store.update_stack(
+            stack_id=default_stack.id, stack_update=stack_update
+        )
+
+
+def test_deleting_default_stack_fails():
+    """Tests that deleting the default stack is prohibited."""
+    client = Client()
+
+    default_stack = client.get_stack(DEFAULT_STACK_NAME)
+    with pytest.raises(IllegalOperationError):
+        client.zen_store.delete_stack(default_stack.id)
+
+
+def test_get_stack_fails_with_nonexistent_stack_id():
+    """Tests getting stack fails with nonexistent stack id."""
+    client = Client()
+    store = client.zen_store
+
+    with pytest.raises(KeyError):
+        store.get_stack(uuid.uuid4())
+
+
+def test_filter_stack_succeeds():
+    """Tests getting stack."""
+    client = Client()
+    store = client.zen_store
+
+    with ComponentContext(
+        c_type=StackComponentType.ORCHESTRATOR, flavor="local", config={}
+    ) as orchestrator:
+        with ComponentContext(
+            c_type=StackComponentType.ARTIFACT_STORE, flavor="local", config={}
+        ) as artifact_store:
+            components = {
+                StackComponentType.ORCHESTRATOR: [orchestrator.id],
+                StackComponentType.ARTIFACT_STORE: [artifact_store.id],
+            }
+            with StackContext(components=components) as stack:
```

Comment:
Wow ! You have unlocked Contextception !

I was about to suggest [contextlib.nested](https://docs.python.org/2/library/contextlib.html#contextlib.nested) to help with nested contexts, but now I see they depend on eachother (the stack does at least).